Linked list basic, application 구현 완료

HyungilNam opened the pull request:
코드 컨밴션에 맞게 변수 및 함수의 이름을 수정하였습니다. 추가적으로
문법의 띄어쓰기 또한 변경하였습니다.

HyungilNam committed:
변수 및 함수 이름을 수정

코드 컨밴션에 맞게 변수 및 함수의 이름을 수정하였습니다. 추가적으로
문법의 띄어쓰기 또한 변경하였습니다.

HyungilNam committed:
연결 리스트를 응용한 학생 정보 프로그램 작성 및 수정

학생의 번호와 이름이라는 정보를 이용해서 삽입, 삭제, 검색, 출력을
수행합니다. 추가적으로 코드 컨밴션에 맞게 함수, 변수의 이름을
수정하였고, 문법의 띄어쓰기도 리펙토링 시에 수정하였습니다.

HyungilNam committed:
Stack 를 이용해 infix를 postfix로변환하는 프로그램 작성

Stack자료구조를 이용해서 infix로 된 수식을 postfix로 변환하는 프로그램을
작성하였습니다.
더불어 코드 컨밴션에 맞게 변수, 함수 등의 이름을 재작성하였고, 문법의
띄어쓰기를 중점적으로 재작성하였습니다.

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+//Linked list
+#include<stdio.h>
```
```c
#include <stdio.h>
```

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+//Linked list
+#include<stdio.h>
+#include<stdlib.h>
```
```c
#include <stdio.h>
```

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+    }   
+}
+
+void DeleteList(ListPointer list_start) {
```
1. 매개변수의 이름은 list로도 충분할 것 같습니다.
2. 만약 리스트에 이미 여러 노드가 달려있다면 그냥 리스트 포인터를 날리는 것만으로 충분할까요?

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+        return 1;
+    }
+
+    ListPointer list_start = malloc(sizeof(struct Node));
```
```c
    ListPointer list = malloc(sizeof(struct Node));
```
리스트 포인터 변수이름으로 list는 어떤가요?
충분히 리스트의 처음을 가리키고 있다는 의미를 내포할 수 있을 것 같아보입니다.

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+
+    while (!feof(fp)) {
+
+    fscanf(fp, "%c %d %d", &input_char, &buffer1, &buffer2);
```
while 루프 안에 전체적으로 indent가 들어가야겠죠?
buffer1, buffer2와 같은 작명보다 더 나은 작명이 있지 않을까요?

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+
+    switch (input_char) {
+      case'i': {
+          if (buffer2 == -1) {
```
굳이 -1 값으로 분기처리를 해야할까요?
insert를 할 때 insert 내부에서 주어진 값을 이용해서 이전 노드를 찾을 때 알아서 처리되지 않을까요?

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+    fscanf(fp, "%c %d %d", &input_char, &buffer1, &buffer2);
+
+    switch (input_char) {
+      case'i': {
```
일관성있게 4 space indent을 사용해야 할 것 같네요.
```c
case 'i':
```

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+          }
+          else {
+              ListPointer list_current_position_i;
+              list_current_position_i = Find(buffer2, list_start);
```
이전 값을 찾는 로직 같아 보이는데, 변수명이 `list_current_position`이 아니라
`previous`정도가 되어야하지 않을까요?
불필요한(내포된 의미의, 중복된) list, position은 굳이 변수 이름에 넣지 않아도 좋을 것 같네요.

그리고 previous node에 대한 탐색 로직이 insert 함수 내부로 들어가야하지 않을까요?

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+    return list_start->next == NULL;
+}
+
+int IsLast(ListPointer list_current_position, ListPointer list_start) {
```
함수 내에서 list_current_position 밖에 사용하지 않는데, 굳이 두 번째 인자를 받을 필요가 있을까요?

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+    return list_current_position->next == NULL;
+}
+
+void Delete(int data, ListPointer list_current_position, ListPointer list_start) {
```
삭제당하는 노드의 이전 노드를 찾는 로직이 함께 있어야 할 것 같네요.

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+void Insert(int data, ListPointer list_start, ListPointer list_current_position) {
+    ListPointer tmp_cell;
+
+    tmp_cell = malloc(sizeof(struct Node));
```
새로운 노드를 만드는 함수를 추상화해서 만들어보세요.
```c
struct Node *CreateNode(int element)
```

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+    ListPointer tmp_cell;
+
+    tmp_cell = malloc(sizeof(struct Node));
+    tmp_cell->element = data;
```
tmp_cell->next 초기화를 안 해주면 null값을 들고 있지 않을텐데 괜찮나요?

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+    tmp_cell = malloc(sizeof(struct Node));
+    tmp_cell->element = data;
+
+    if (IsLast(list_start,list_current_position)) {
```
의미 없는 분기입니다.
마지막이든 아니든 똑같은 로직으로 삽입이 이루어지지 않을까요?

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+}
+
+ListPointer FindPrevious(int data, ListPointer list_start) {
+    
```
no whitespace

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+    }
+}
+
+ListPointer FindPrevious(int data, ListPointer list_start) {
```
일관성있게 변수명을 작성해야합니다.
현재 struct Node에는 data가 element라는 이름으로 들어있는데, 여기서는 data라는 이름으로 받고 있네요.
한 가지로 통일되어야 하지 않을까요?

kwangilcho commented on linkedlist.c:
```diff
@@ -0,0 +1,166 @@
+
+      case'd': { 
+          ListPointer list_current_position_d;
+          list_current_position_d = FindPrevious(buffer1, list_start);
```
delete 함수 내부로 들어가야할 것 같습니다.

kwangilcho reviewed (commented):
코드 짜느라 고생하셨습니다.
코멘트를 기반으로 전체적으로 다시 작성해야할 것 같습니다.

kwangilcho commented:
변수 이름에 타입을 지칭하는 단어를 굳이 쓸 필요가 없습니다.
C 언어는 정적 타입 언어이기 때문에 반드시 변수를 생성할 때 어떤 타입의 변수인지를 명시하게 되어있습니다.
즉, C로 코드를 작성하는 사람은 반드시 타입을 보게 되어있다는 뜻입니다.
그러므로 변수에 타입에 대한 정보를 담는 것 자체가 중복이라 볼 수 있습니다.

HyungilNam committed:
문법 및 구조 수정

변경사항1
1. include 띄어쓰기 및 while, switch indent 수정하였습니다.
2. 매개변수 list 변경하였고, input_char, buffer1,2 를 command, target, position으로, list_current_position을 각 함수 상황에  맞게 변경하였습니다.
3. DeleteList는 해당 리스트에 있는 노드들을 모두 삭제하기 때문에 순차적
삭제가 가능하여 리스트 포인터만 날려도 가능할 것으로 보입니다.
4. Insert와 DeleteNode 모두 해당 함수에서 위치를 찾아서 삽입하도록
수정하였습니다.
IsLast 함수 인자 수정하였습니다.

HyungilNam committed:
추가적인 문법과 구조를 수정

변경사항2
1. 새로운 노드를 만드는 함수로 Insert의 tmp_cell을 수정하였습니다. 또한
MakeEmpty를 초기에 한번 해줄 때 기능이 중복되어서 삭제하였습니다.
2. Insert 함수에 있던 의미 없는 분기 수정하였습니다.
3. 구조체와 함수에 있는 변수명 통일하였습니다.

HyungilNam committed:
linkedlist.c 파일의 수정 사항을 반영한 새로운 파일 작성

linkedlist.c의 변경사항을 반영하여 다시 push한 완성본입니다.

JunsuLime commented on infix_to_postfix.c:
```diff
@@ -0,0 +1,125 @@
+
+char target_line[DATACHAR];
+
+struct Stack{
```
안녕하세요 광일공방 자료구조 프로젝트에 들어오게 된 임준수라고 해요 :)
서로 리뷰같이 해주면 좋을 것 같아서 코드리뷰 남깁니당!

`struct Stack{` 을 코딩 컨벤션에 맞게 `struct Stack {` 로 하는 것이 더 좋다고 생각해요.

코딩컨벤션에 맞는 코드를 커밋하기 위해서는 IDE 나 [linter](https://ko.wikipedia.org/wiki/%EB%A6%B0%ED%8A%B8_(%EC%86%8C%ED%94%84%ED%8A%B8%EC%9B%A8%EC%96%B4)) 의 도움을 사용하는 것도 괜찮다고 생각해요.
저는 현재 C 작업할 때 Visual Code 를 사용하고 `ms-vscode.cpptools` 라는 플러그인을 붙여서 사용하고 있어요.

구글 컨벤션에 맞되 들여쓰기를 4 space 로 하기위해서 
```json
{
    "files.associations": {
        "*.c": "c",
    },
    "C_Cpp.clang_format_fallbackStyle": "{ BasedOnStyle: Google, IndentWidth: 4 }",
}
```
Visual Code User Settings 에서 다음과 같은 설정을 해주었어요.

이렇게 설정한 후 자동줄맞춤 기능을 이용하면 코딩컨벤션을 실수로 맞추지 못한 부분에 대해 잡을 수 있습니당

현길선생님께서 어떤 툴을 쓰시는지는 모르겠지만 대부분 IDE 에는 이와같은 기능이 있을거에요!

HyungilNam replied:
제가 우분투를 쓰는데 거기 까지는 몰랐네요 ㅠ

JunsuLime commented on infix_to_postfix.c:
```diff
@@ -0,0 +1,125 @@
+    int max_stack_size;
+};
+
+void MakeEmpty(struct Stack* S) {
```
https://google.github.io/styleguide/cppguide.html#Variable_Names
https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

변수의 이름은 모두 소문자 혹은 언더바를 섞은 소문자를 사용하는 것을 원칙으로 해용 ㅎㅎ
변수명은 단어의 생략을 피하고 의미있는 단어로 구성하는 것이 좋을 것 같아요!

위의 링크는 이에 대해 자세히 설명 된 구글 컨벤션 페이지 링크입니당 :)

JunsuLime commented on infix_to_postfix.c:
```diff
@@ -0,0 +1,125 @@
+
+void MakeEmpty(struct Stack* S) {
+    S->top = -1;
+    S->max_stack_size = 40;
```
1. stack 의 크기가 항상 40으로 고정되어있는 경우라면
    40 이라는 상수는 `#define` 으로 빼는 것이 가독성 측면에서 더 좋을 것 같아요.

2. 아니면 사용자가 max 값을 인자로 입력 할 수 있게 하는 것도 괜찮다고 생각해요.
    현재 과제가 infix to postfix 인데 데이터 길이와 경우에 따라 size 40은 터질 수 있다고 생각해요. 따라서 `MakeEmpty(struct *stack, int max_size)` 로 바꾸는 것이 여러 상황에 대해 유연하게 대처 할 수있는 방법인 것 같아요!

JunsuLime commented on infix_to_postfix.c:
```diff
@@ -0,0 +1,125 @@
+//우선순위 정하기
+int Priority(char operator){
+    switch (operator) {
+      case '(': case')': 
```
https://google.github.io/styleguide/cppguide.html#Loops_and_Switch_Statements

구글 컨벤션에서는 switch 문을 쓸 때 다음과 같이 쓰더라고요 ㅎㅎㅎㅎ 위에 링크 올려놨어요 !

JunsuLime commented on infix_to_postfix.c:
```diff
@@ -0,0 +1,125 @@
+}
+
+void Push(char word, struct Stack* S) {
+    S->top = S->top + 1;
```
52 번째 줄 `stack->top--` 처럼 여기도 `stack->top++` 을 해보는 것은 어떨까요?

JunsuLime commented on infix_to_postfix.c:
```diff
@@ -0,0 +1,125 @@
+                //스택의 탑에 있는 문자 우선순위가 들어오는 문자의 우선순위보다 낮은 경우, 푸쉬
+                if (p > Priority(S->key[S->top])) {
+                    //test
+                    //printf("들어갈놈%c: %d, 스택안에 있는놈%c: %d\n",targetLine[i],p,S->key[S->top],Priority(S->key[S->top]));
```
커밋을 할 때 코드 설명에 대한 주석말고 88번째 줄과 같이 안쓰는 코드를 지우기 위한 주석은 안 들어가는 것이 좋을 것 같아요!

JunsuLime commented on infix_to_postfix.c:
```diff
@@ -0,0 +1,125 @@
+        //피연산자와 연산자 구분 및 처리
+        if (isdigit(target_line[i])) {
+            strncat(converted_line, &target_line[i], 1);
+        }
```
if-else 문으로 인해 indent 가 깊어지는 문제를 해결해 볼 수 있을 것 같아요!

indentation 이 깊어지면 코드 가독성이 떨어지는 문제가 발생할 수 있어요.

현재 코드에서 else 를 없애고
```c
if (isdigit(target_line[i])) {
    strncat(converted_line, &target_line[i], 1);
    continue;
}
```
이렇게 한다면 1. indentation 도 줄고, 2. 머리 속으로 테스트 할 때도 if 문에 걸렸을 때 바로 continue 를 보아서 뒷 글을 안읽고 빠르고 루프를 돌 수 있을거에요!

이와 같은 부분이 몇개 있어서 한번 고민해보는 것도 좋다고 생각합니다.

JunsuLime commented on infix_to_postfix.c:
```diff
@@ -0,0 +1,125 @@
+            
+            //스택만들고 초기화 설정
+            struct Stack* S;
+            S = malloc(sizeof(struct Stack));
```
혹시 이 줄이 컴파일 시 경고가 뜨지않나요? 혹여 에러가 안 뜨더라도
명시적으로 `(Stack*)malloc(struct Stack);` 형변환을 해주는 것이 좋아용 ㅎㅎ

JunsuLime commented on linkedlist _modified.c:
```diff
@@ -0,0 +1,153 @@
+
+    if (fp == NULL) {    
+        printf("Input Error");
+        return 1;
```
<stdlib.h> 에는 `EXIT_FAILURE` 가 있어요 ㅎㅎㅎㅎ
그래서 비정상 종료 시 

1. `return EXIT_FAILURE` 혹은
2. `exit(EXIT_FAILURE)` 을 호출 하더라고요 !

참고로 EXIT_FAILURE 는 1 입니당ㅎㅎㅎㅎ

다른 부분도 비슷한 코드가 있는데 고민해 보는 것은 어떨까요?

JunsuLime commented on linkedlist _modified.c:
```diff
@@ -0,0 +1,153 @@
+            }
+            default:
+                continue;
+        }
```
Insert 나 Delete 처럼  Find 와 Print 도 함수로 빼보는 것은 어떨까요?
밑의 LinkedList 도 같은 방법으로 함수로 빼봐요 ㅎㅎ

JunsuLime commented on linkedlist.c:
```diff
@@ -0,0 +1,153 @@
+typedef struct Node* ListPointer;
+
+struct Node {
+    int target;
```
Node 내의 item 을 담는 용도로 쓰이는 변수로 보이는데 `item`  혹은 `element` 와 같은 이름으로 바꾸는 건 어떨까요?

JunsuLime commented on linkedlist.c:
```diff
@@ -0,0 +1,153 @@
+ListPointer Find(int target, ListPointer list);
+void Insert(int target, int position, ListPointer list);
+void DeleteList(ListPointer list);
+
```
ListPointer 를 이용해서 LinkedList 의 모든 동작을 가능하게 할 수 있지만, 제 3자가 코드를 읽을 때 혹은 해당 코드를 재사용하려고 할 때는 사용하고 힘들 것 같아요.

함수를 제공해준다면 내부 구현은 감추고 인터페이스만 제공해도 문제 없이 돌아가는 것이 최선이라고 생각해요. 이런 관점에서 볼 때 ListPointer 또한 LinkedList 의 내부구현(implementation) 이라 보입니다.

이렇게 하기보다

```
struct LinkedList { 
    int element;
    struct Node *next;
}
```
이런 식으로 Node 를 감추어 
LinkedList 를 다른 사람이 사용할 때 보여지는 것이

1. LinkedList 구조체
2. 함수들 

만 있게 되면 코드의 가독성이 더 높아질 거라 생각 되요!!

JunsuLime commented on linkedlist.c:
```diff
@@ -0,0 +1,153 @@
+    while (!feof(fp)) {
+        fscanf(fp, "%c %d %d", &command, &target, &position);
+        switch (command) {
+            case'i': {
```
https://google.github.io/styleguide/cppguide.html#Loops_and_Switch_Statements

switch 문의 코딩컨벤션을 확인해보아요 :)

JunsuLime commented on linkedlist.c:
```diff
@@ -0,0 +1,153 @@
+}
+
+ListPointer CreateNode(int target) {
+    ListPointer new_node = malloc(sizeof(struct Node));
```
malloc 함수는 void* 보이드 형 포인터를 반환해줍니다.
이를 ListPointer 변수에 할당 할 때는 경고가 뜨지 않더라도 명시적으로 형변환을 해주는 것이 좋아요!

JunsuLime commented on linkedlist.c:
```diff
@@ -0,0 +1,153 @@
+    previous = FindPrevious(target, list);    
+    if (previous->next == NULL) {
+        printf("Deletion failed : element %d is not int the list \n", target);
+    }   
```
if 문 이후 함수내에 실행 할 코드가 없다면 코드의 간결함을 위해

```c
if (previous->next == NULL) {
    printf("hello world");
    return;
}
// else 없이 이후 코드
```
이런 방식으로 하는게 좋습니당 :)

다른 부분에도 이와 같은 경우가 몇개 있는데 찾아서 적용해봐요 ㅎㅎㅎㅎ

JunsuLime commented on linkedlist.c:
```diff
@@ -0,0 +1,153 @@
+
+ListPointer FindPrevious(int target, ListPointer list) {
+    ListPointer previous = list;
+    while (previous->next != NULL && previous->next->target != target)
```
`if (previous->next != NULL)` 은 `if (previous->next)` 와 같은 코드에요.

사실 취향 타는 부분인 것 같아서 ㅎㅎㅎㅎ 참고로 올립니당 ㅎㅎㅎㅎ

JunsuLime commented on linkedlist _modified.c:
```diff
@@ -0,0 +1,153 @@
+void Insert(int target, int position, ListPointer list) {
+    ListPointer after_target, tmp_cell;
+    tmp_cell = CreateNode(target);
+    if (position == -1) {
```
-1 은 #define 으로 주면 더 좋을 것 같아요! 
상수 값은 #define 으로 주어 코드의 가독성을 높여보아요 ㅎㅎㅎㅎ

JunsuLime commented on linkedlist.c:
```diff
@@ -0,0 +1,153 @@
+void Insert(int target, int position, ListPointer list) {
+    ListPointer after_target, tmp_cell;
+    tmp_cell = CreateNode(target);
+    if (position == -1) {
```
상수값은 #define 으로 빼보도록 해요 :)

JunsuLime commented on linkedlist_application.c:
```diff
@@ -0,0 +1,239 @@
+    enum command_t command;
+    char student_id[MAX_ARGUMENT];     
+    char student_name[MAX_ENROLLMENT];
+};
```
https://google.github.io/styleguide/cppguide.html#Class_Format
https://google.github.io/styleguide/cppguide.html#Class_Format

클래스 이름과 enumerator 이름을 컨벤션에 맞춰봐요

HyungilNam replied:
여기는 어떤 부분을 수정해야할지 잘 모르겠어요 ㅠ

HyungilNam replied:
음 어떻게 수정할지 감이 잘 안오네요 ㅠ

JunsuLime replied:
Enum 과 struct 의 이름은 대문자로 시작하는 카멜케이스를 사용하시면 돼요!

ex)
enum Command

HyungilNam replied:
아 넵넵 ㅎ

JunsuLime commented on linkedlist_application.c:
```diff
@@ -0,0 +1,239 @@
+                ParseCommandLine(&commandline);
+                switch (commandline.command) {
+                  case INSERT:
+                      {
```
불필요한 중괄호의 사용이라고 생각되요!

JunsuLime commented on linkedlist_application.c:
```diff
@@ -0,0 +1,239 @@
+                               break;
+                           }
+                           else {
+                           //Insertion function
```
else 문이 없어도 괜찮아요 ㅎㅎㅎㅎ

JunsuLime commented on linkedlist_application.c:
```diff
@@ -0,0 +1,239 @@
+                           printf("Insertion Success : %s \n", commandline.student_id);
+                           printf("Current List>");
+                           PrintNode(L);
+                           printf("\n");
```
PrintNode 뒤에 항상 printf("\n") 이 붙으니 이를 PrintNode 안으로 넣어보는 것은 어떨까요?

JunsuLime commented on linkedlist_application.c:
```diff
@@ -0,0 +1,239 @@
+
+void MakeEmpty(struct Node* L) {
+    L->given_id;
+    L->given_name;
```
아무것도 넣어주는 것이 아니라면 154 와 155 는 없어도 괜찮습니다.

JunsuLime commented on linkedlist_application.c:
```diff
@@ -0,0 +1,239 @@
+struct Node* Find(char* student_id, struct Node* L) {
+    struct Node* search = L->next;
+
+    int s_given_id = 1;
```
s_given_id 가 1로 초기화 되어있는데 이는 사용되지 않는 값이에요 ㅎㅎㅎㅎ
1을 대입하지않고 선언만 해주는 것으로 바꿔봐요!

그리고 변수명을 더 명확하게 짓는 것이 좋아보여요 :)

HyungilNam committed:
infix_to_postfix.c 피드백 반영 후 수정

Stack이라는 구조체 선언시 띄어쓰기 반영하였고, 변수명 코딩 컨밴션 반영하였습니다. 그리고 하드코딩되어있는 stack size를 #define으로 따로 빼내었습니다. switch문, ++연산자, 동적할당 시 형변환 수정했고, 불필요한 주석 제거 및 if-else 문이 깊어지지 않도록 수정했습니다. <stdlib.h>에 비정상 종료 예외처리 반영하였습니다.

HyungilNam committed:
linkedlist_modified.c를 재작성

원래 linkedlist.c파일의 내용을 한번 피드백 받고, linkedlist
_modified.c파일을 새로 만들었던 것인데 피드백에 혼선이 있던 것으로 보여
두 파일의 피드백을 모두 반영하여 linkedlist_modified.c로 수정하여 로드
하였습니다. main 함수에 있던 Print와 Find를 함수로 빼고, 새부 수정
사항들을 반영하였습니다.

HyungilNam committed:
linkedlist_application.c파일 피드백 반영하여 수정

enum과 struct 명 선언시 대문자로 수정하였습니다. 불필요한 중괄호 ,
else문, 내용을 제거하여 수정하였습니다. 추가적으로 s_given_id 와 같은 모호한
변수명을 tmp_given_id로 수정하였습니다.

JunsuLime reviewed (approved):
고생많으셨어요! 머지할게요 ㅎㅎㅎㅎ

kwangilcho commented:
고생많으셨습니다 👍
